### PR TITLE
feat: Search facets

### PR DESCRIPTION
**Pull request by alexgarel**
* introduce searchalicious-facets and searchalicious-facet-terms components to:
  * display facets
  * add filters to search
* refactor base of results handling in the search-results-ctl mixin
* change API for facets to be more easy to use:
  * add a facets parameter
  * return easier to parse facets results

Limitations:
* categories_tags seems not to work (but it's a backend problem)
* we need the API to send back selected value because they may disapear while crossed with full text, so user won't be able to uncheck them ! (but first backend must support it)
* for now only the "terms" api is supported, and facet is always sorted by count (while it could be sorted by name for nutri-score for example)

https://github.com/openfoodfacts/search-a-licious/assets/144788/824e5a3d-faeb-473a-b02b-b779cc9ce26e

### CODE REVIEW

File: app/_types.py [hunk 1]
```diff
@@ -7,6 +7,29 @@
 JSONType = dict[str, Any]
 
 
+class FacetItem(BaseModel):
+    """Describe an entry of a facet"""
```
Comment (teolemon) [inline app/_types.py:11]:
```suggestion
    """Describes an entry of a facet"""
```

File: app/_types.py [hunk 2]
```diff
@@ -7,6 +7,29 @@
...
+
+    key: str
+    name: str
+    count: int
+    """The number of elements for this value"""
+
+
+class FacetInfo(BaseModel):
+    """Search result for facet"""
```
Comment (teolemon) [inline app/_types.py:20]:
```suggestion
    """Search result for a facet"""
```

File: README.md [hunk 3]
```diff
@@ -57,6 +57,16 @@ uvicorn app.api:app --reload --port=8001 --workers=4
 ```
 Note that it's important to use port 8001, as port 8000 will be used by the docker version of the search service.
 
+
+
+To debug the backend app:
+* stop api instance: `docker compose stop api`
```
Comment (teolemon) [inline README.md:63]:
```suggestion
* stop API instance: `docker compose stop api`
```

**Comment by stephanegigandet** [comment]
Cool! It would be nice to have a "Apply new filters" button that appears below the facets, as it's not intuitive that we need to click on the search button again to apply them.

File: data/config/openfoodfacts.yml [hunk 4]
```diff
@@ -123,6 +123,16 @@ indices:
...
+    facets:
+      # include all fields
+      default:
+        order:
+        - categories_tags
+        - brands_tags
+        - labels_tags
+        - ecoscore_grade
+        - nova_groups
+        - ecoscore_grade
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply data/config/openfoodfacts.yml:135]: I have to remove that, this is not useful any more… now it's the interface that requests facets (more logical).

**Comment by alexgarel** [comment]
> It would be nice to have a "Apply new filters" button that appears below the facets, as it's not intuitive that we need to click on the search button again to apply them.

Yes, in another PR ;-)

File: frontend/src/search-facets.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  searchFilter(): string | undefined {
+    throw new Error('renderFacet not implemented: implement in sub class');
+  }
+
+  override render() {
+    if (this.infos) {
+      return this.renderFacet();
+    } else {
+      return html``; // FIXME: is this ok ?
+    }
```
Comment (Valimp) [inline frontend/src/search-facets.ts:126]: What this comment highlight ? 
If the return html'' render nothing ?

**Review by Valimp — COMMENTED** [review]
Looks good for me

**Comment by Valimp** [comment]
> > It would be nice to have a "Apply new filters" button that appears below the facets, as it's not intuitive that we need to click on the search button again to apply them.
> 
> Yes, in another PR ;-)

It's possible to updates results when you check any filter

File: frontend/src/search-facets.ts (hunk 5 above)
Comment (alexgarel) [reply frontend/src/search-facets.ts:126]: Yes, I was not sure when I code, now it seems ok, I will remove the FIXME :-)

**Comment by alexgarel** [comment]
> > > It would be nice to have a "Apply new filters" button that appears below the facets, as it's not intuitive that we need to click on the search button again to apply them.
> > 
> > 
> > Yes, in another PR ;-)
> 
> It's possible to updates results when you check any filter

I personally don't like it much because it makes hard to select two filters at once…

**Review by Valimp — APPROVED** [review]

**Pull request merged**